Regras de Associação: ATIVIDADE 2
Aluna: Mayara Cardoso Simões

2) Utilize a base de dados maior (mercado2) que apresenta o histórico de compras de um supermercado francês em 1 semana. Teste o seu algoritmo e responda as seguintes perguntas:

In [72]:
import pandas as pd
from apyori import apriori

base_mercado = pd.read_csv('data/mercado2.csv', header = None)
transacoes = []
qtd_colunas = base_mercado.shape[1]
for i in range(0, len(base_mercado)):
     transacoes.append([str(base_mercado.values[i, j]) for j in range(qtd_colunas) if str(base_mercado.values[i, j]) != 'nan'])

a) Vamos supor que você quer extrair Regras da Associação para os produtos que são vendidos pelo menos quatro vezes ao dia, ou seja, 28 vezes por semana (4*7). Adicione o valor de suporte de acordo com o problema.

Para adicionar o valor de suporte, você deve calcular o valor de 28 (quantidade de produtos vendidos) dividido pela quantidade total de produtos na base de dados.

Execute o algoritmo e responda qual é o resultado.

In [73]:
suporte_minimo = 28 / len(transacoes)

print(f"Suporte mínimo calculado: {suporte_minimo:.4f}")

regras = apriori(transacoes, min_support=suporte_minimo)
resultados = list(regras)
print(f"Total de regras geradas: {len(resultados)}")

Suporte mínimo calculado: 0.0037
Total de regras geradas: 1104


Resultado:
Suporte mínimo calculado: 0.0037
Total de regras geradas: 1104

b) Adicione o valor de confiança para 0.2 e o lift para 3. Quantos registros foram retornados?

In [74]:
regras = apriori(transacoes, min_support=suporte_minimo, min_confidence=0.2, min_lift= 3)
resultados = list(regras)
print(f"Total de regras geradas: {len(resultados)}")

Total de regras geradas: 44


Total de regras geradas: 44

c) Adicione o código abaixo para poder visualizar melhor os dados. Faça comentário do que acontece em cada linha do código.

In [75]:
A = []
B = []
suporte = []
confianca = []
lift = []

# Percorre todas as regras encontradas
for resultado in resultados:
    s = resultado[1]            # Valor do suporte da regra
    result_rules = resultado[2] # Lista com as relações entre itens (se → então)

    # Percorre regra dentro do resultado
    for result_rule in result_rules:
        a = list(result_rule[0])   # Se (parte A da regra)
        b = list(result_rule[1])   # Então (parte B da regra)
        c = result_rule[2]         # Confiança
        l = result_rule[3]         # Lift

        # Armazena cada valor nas listas
        A.append(a)
        B.append(b)
        suporte.append(s)
        confianca.append(c)
        lift.append(l)

# Cria um DataFrame com todas as informações das regras
rules_df = pd.DataFrame({
    'A': A,
    'B': B,
    'suporte': suporte,
    'confianca': confianca,
    'lift': lift
})

# Ordena as regras pelo valor de lift
rules_df.sort_values(by='lift', ascending=False)

A  \
40               [mineral water, whole wheat pasta]   
0                                     [light cream]   
2                                           [pasta]   
50                 [ground beef, frozen vegetables]   
6                                           [pasta]   
11                       [chocolate, herb & pepper]   
54                 [ground beef, frozen vegetables]   
15                              [eggs, ground beef]   
5                               [whole wheat pasta]   
33                       [herb & pepper, spaghetti]   
32                   [mineral water, herb & pepper]   
47                           [olive oil, chocolate]   
21                        [soup, frozen vegetables]   
4                                    [tomato sauce]   
44                              [eggs, ground beef]   
1                            [mushroom cream sauce]   
28                               [tomatoes, shrimp]   
55    [mineral water, spaghetti, frozen vegetables]   
48            [mineral water, chocolate, spaghetti]   
53               [mineral water, milk, ground beef]   
42                            [olive oil, tomatoes]   
26                      [shrimp, frozen vegetables]   
38                   [spaghetti, whole wheat pasta]   
29                   [spaghetti, frozen vegetables]   
27                    [tomatoes, frozen vegetables]   
52         [mineral water, milk, frozen vegetables]   
39                            [mineral water, soup]   
34                              [milk, ground beef]   
16                            [herb & pepper, eggs]   
57                              [milk, ground beef]   
30                            [tomatoes, spaghetti]   
3                                   [herb & pepper]   
31                       [grated cheese, spaghetti]   
13                       [cooking oil, ground beef]   
20                   [olive oil, frozen vegetables]   
9                    [chocolate, frozen vegetables]   
7                                      [milk, cake]   
23                          [mineral water, shrimp]   
43                            [tomatoes, spaghetti]   
49         [mineral water, eggs, frozen vegetables]   
46    [mineral water, chocolate, frozen vegetables]   
35                              [soup, ground beef]   
18                   [spaghetti, frozen vegetables]   
8                            [chicken, ground beef]   
51  [mineral water, ground beef, frozen vegetables]   
24                   [spaghetti, frozen vegetables]   
10                              [chocolate, shrimp]   
22                                 [milk, tomatoes]   
19                        [milk, frozen vegetables]   
37                                [milk, spaghetti]   
14                         [cooking oil, spaghetti]   
41                            [spaghetti, pancakes]   
12                                [chocolate, soup]   
17                                 [red wine, eggs]   
45                 [mineral water, chocolate, eggs]   
56                 [mineral water, milk, spaghetti]   
25                   [spaghetti, frozen vegetables]   
36                            [ground beef, shrimp]   

                             B   suporte  confianca      lift  
40                 [olive oil]  0.003866   0.402778  6.115863  
0                    [chicken]  0.004533   0.290598  4.843951  
2                   [escalope]  0.005866   0.372881  4.700812  
50       [mineral water, milk]  0.003733   0.220472  4.593788  
6                     [shrimp]  0.005066   0.322034  4.506672  
11               [ground beef]  0.003999   0.441176  4.490183  
54  [mineral water, spaghetti]  0.004399   0.259843  4.350622  
15             [herb & pepper]  0.004133   0.206667  4.178455  
5                  [olive oil]  0.007999   0.271493  4.122410  
33               [ground beef]  0.006399   0.393443  4.004360  
32               [ground beef]  0.006666   0.390625  3.975683  
47  [mineral water, spaghetti]  0.00

d) Ordene pelo valor de confiança para verificar qual regra tem maior confiabilidade. Qual o valor da regra com maior confiança?

In [76]:
rules_df.sort_values(by='confianca', ascending=False)

A  \
42                            [olive oil, tomatoes]   
13                       [cooking oil, ground beef]   
17                                 [red wine, eggs]   
36                            [ground beef, shrimp]   
21                        [soup, frozen vegetables]   
38                   [spaghetti, whole wheat pasta]   
11                       [chocolate, herb & pepper]   
20                   [olive oil, frozen vegetables]   
49         [mineral water, eggs, frozen vegetables]   
35                              [soup, ground beef]   
46    [mineral water, chocolate, frozen vegetables]   
8                            [chicken, ground beef]   
51  [mineral water, ground beef, frozen vegetables]   
40               [mineral water, whole wheat pasta]   
12                                [chocolate, soup]   
33                       [herb & pepper, spaghetti]   
32                   [mineral water, herb & pepper]   
4                                    [tomato sauce]   
2                                           [pasta]   
55    [mineral water, spaghetti, frozen vegetables]   
28                               [tomatoes, shrimp]   
53               [mineral water, milk, ground beef]   
52         [mineral water, milk, frozen vegetables]   
16                            [herb & pepper, eggs]   
3                                   [herb & pepper]   
31                       [grated cheese, spaghetti]   
6                                           [pasta]   
30                            [tomatoes, spaghetti]   
18                   [spaghetti, frozen vegetables]   
23                          [mineral water, shrimp]   
14                         [cooking oil, spaghetti]   
1                            [mushroom cream sauce]   
45                 [mineral water, chocolate, eggs]   
10                              [chocolate, shrimp]   
22                                 [milk, tomatoes]   
0                                     [light cream]   
56                 [mineral water, milk, spaghetti]   
7                                      [milk, cake]   
5                               [whole wheat pasta]   
54                 [ground beef, frozen vegetables]   
27                    [tomatoes, frozen vegetables]   
48            [mineral water, chocolate, spaghetti]   
26                      [shrimp, frozen vegetables]   
29                   [spaghetti, frozen vegetables]   
47                           [olive oil, chocolate]   
9                    [chocolate, frozen vegetables]   
39                            [mineral water, soup]   
34                              [milk, ground beef]   
50                 [ground beef, frozen vegetables]   
25                   [spaghetti, frozen vegetables]   
43                            [tomatoes, spaghetti]   
15                              [eggs, ground beef]   
24                   [spaghetti, frozen vegetables]   
19                        [milk, frozen vegetables]   
37                                [milk, spaghetti]   
41                            [spaghetti, pancakes]   
44                              [eggs, ground beef]   
57                              [milk, ground beef]   

                             B   suporte  confianca      lift  
42                 [spaghetti]  0.004399   0.611111  3.509912  
13                 [spaghetti]  0.004799   0.571429  3.281995  
17                 [spaghetti]  0.003733   0.528302  3.034297  
36                 [spaghetti]  0.005999   0.523256  3.005315  
21                      [milk]  0.003999   0.500000  3.858539  
38                      [milk]  0.003999   0.454545  3.507763  
11               [ground beef]  0.003999   0.441176  4.490183  
20                      [milk]  0.004799   0.423529  3.268410  
49                      [milk]  0.003733   0.411765  3.177620  
35                      [milk]  0.003999   0.410959  3.171402  
46                      [milk]  0.003999   0.410959  3.171402  
8                       [milk]  0.00

Se comprar azeite de oliva e tomate, então há uma probabilidade de 61,11% de também comprar espaguete.